### ISSUE TITLE
Update to Mongo 3.2

### ISSUE THREAD

**Issue by tmeasday** [meteor/meteor#6957]
### PR: https://github.com/meteor/meteor/pull/7163

This issue is an attempt to outline what needs to be considered as part of this transition. We [plan](https://github.com/meteor/meteor/blob/devel/Roadmap.md#mongodb-updates) to look at this in the short/medium term, however if someone is keen to contribute to the project and wants to see it land earlier, this would be a great opportunity to make a larger contribution!

### Update Mongo Driver to 3.2 compatible version https://github.com/meteor/meteor/issues/5763

 - [x] We need to update the version of the `node-mongodb-native` package used here: https://github.com/meteor/meteor/blob/devel/packages/npm-mongo/package.js to at least 2.1.0 (I guess we may as well get the latest 2.1.18).

 - [x] Check our uses of the driver in https://github.com/meteor/meteor/blob/devel/packages/mongo/ and other core code. Supposedly the API is backwards compatible, although there are some small issues mentioned here: https://github.com/meteor/meteor/issues/5763#issuecomment-163055637

 - [x] Ensure tests pass

(We should also probably remove the `npm-mongo` wrapper package and just peer-depend on the npm package from `mongo`, but I think we should wait to do this, it'll be a larger issue we have to solve with the 1.4 move to npm).

### Update bundled Mongo in the dev bundle to 3.2 https://github.com/meteor/meteor/issues/5809

 - [x] Update our fork of Mongo [https://github.com/meteor/mongo] to 3.2 (there has been work done already in this direction: https://github.com/meteor/mongo/tree/ssl-r3.2.0). This fork allows us to statically link Mongo. (**Note we didn't actually do this, using vanilla**)
 
 - [x] Update https://github.com/meteor/meteor/blob/devel/scripts/build-mongo-for-dev-bundle.sh to build + include Mongo 3.2. (**Note we didn't actually do this, using vanilla**)

 - [x] Update generate-dev-bundle scripts

 - [x] Ensure that if you run the 3.2 Mongo server against an (older) 2.6 local database, it gives a meaningful error that suggests that you `meteor reset` or look into taking steps to upgrading your db contents

 - [x] Figure out if we need to turn journalling off or if there's a way to get the minimum journal size down.

### Errata

 - [x] Update https://github.com/meteor/meteor/blob/devel/scripts/admin/test-packages-with-mongo-versions.rb to also test 3.2

 - [x] Ensure above tests pass on Mongo 2.4 or decide to drop support for it.

 - [x] Document changed behavior of update return value, as per https://github.com/meteor/meteor/issues/6957#issuecomment-222633226

 - [ ] Check if instructions for setting up oplog need updating
 
### Notes on testing mongo binary SSL support:

```
cd /tmp
openssl req -newkey rsa:2048 -new -x509 -days 365 -nodes -out mongodb-cert.crt -keyout mongodb-cert.key
cat mongodb-cert.key mongodb-cert.crt > mongodb.pem
./mongod --sslMode requireSSL --sslPEMKeyFile /tmp/mongodb.pem  --dbpath /tmp/db
./mongo --ssl localhost/test --sslCAFile /tmp/mongodb-cert.crt
```
### I don't think we need to do this:

 - [ ] Resolve building the binary on developer machines - this means continue to maintain our fork of `node-mongodb-native` to include this commit https://github.com/mongodb/node-mongodb-native/commit/634759e6326dc19a228df66ddb309285532f3b8a (which would now need to be against https://github.com/christkv/mongodb-core/blob/1.3/package.json)

**Comment by Fabs** [comment]
Hi @tmeasday. I am in between jobs right now, and I've been looking to contribute to meteor for a while. I think this is a great opportunity to start with, and something I can tackle. 

I have a couple questions, since I've never contributed before.
- Is there any contribution guide you are used to point new contributors to? 
- Which channel would be the most appropriate to talk with the people that worked with this part of meteor before? (in case I need help or help deciding an approach to solve something)

Anyways, I will start digging deeper on this issue later today, and if anyone eager to work with me on this shows up, I would be grateful to work together.

**Comment by tmeasday** [comment]
Hi @Fabs! Thanks for volunteering.

We have some basic contributing guidelines here: https://github.com/meteor/meteor/blob/devel/Contributing.md#making-changes-to-meteor-core

I think for now we can just communicate here on this issue, I'll do my best to be responsive. 

I'm not sure if you had a plan around how you wanted to tackle this, but I've created a `mongo-3.2` branch -- a start would be to a PR to it that:
 1. Updates the `npm-mongo` package to just use the vanilla `node-mongodb-native@2.1.9` package
 2. Passes all `test-packages` tests when run with `MONGO_URL` and `MONGO_OPLOG_URL` set to point to a 3.2 server (for now just run Mongo locally on your machine by hand).

We'll still need to evaluate if we need a fork of that package or not (with the overarching concern being: can you install it without a toolchain on our 4 platforms), but this seems like a nice useful self container piece of work.

**Comment by dgtlife** [comment]
I have made modifications to `mongo` and `npm-mongo` that involve the use of the latest MongoDB Node.js Driver (albeit still in the `npm-mongo` wrapper) and the addition of support for SSL, and X.509 Auth on SSL via additional environment variables. This is a quick hack to keep me on track while I wait for the "official" package(s) to be made available. You can find the mods at [https://github.com/dgtlife/meteor-mongo-for-3.2.x](url).

Hopefully this helps the effort.

**Comment by Fabs** [comment]
Nice :). @tmeasday, I set up an environment with mongo `2.6.7` (from `dev_build`) and `3.2.6` (from homebrew) both with 3 replicas oplog and a single daemon. For the record, I have a failing test `httpcall - S: errors  true - not true - asyncBlock 0 (2 times)`. It happens for all setups except `3.2.6 with oplog`, this one does not even connect with the current driver. 

Anyways, going forward and changing the driver I have a few questions:
1. I made the switch on Npm.depends at `npm-mongo` to mongodb `2.1.18` (most recent on npm). Was  there any particular to cite `2.1.9`? 
2. I forked meteor and merged `devel` onto `mongo-3.2` there was a couple small changes. I plan to keep merging devel frequently so I don't end up much behind. Do you think this is a good strategy?

Anyway, I think I am set up and I can move forward with the work to fix the errors that I have, now that I changed to the new driver. I will keep you guys updated. @dgtlife, thanks for the link, I took a brief look and I am quite sure I will end up using it for help. 

So, the journey begins :)

**Comment by Fabs** [comment]
Also, I though it was not always, so I did not mention. But very frequently (because one or two times i did not) I also get the error `accounts C: reconnect auto-login -  fail — fail - message Issue #4970 has occured.`. I looked #4970 and its closed.

**Comment by tmeasday** [comment]
Cool!

> For the record, I have a failing test

A failing test in a vanilla setup? That would be quite surprising. Or do you mean failing when you run mongo yourself (from the dev bundle) and use `MONGO_URL` and `MONGO_OPLOG_URL`?

> I made the switch on Npm.depends at npm-mongo to mongodb 2.1.18 (most recent on npm). Was there any particular to cite 2.1.9? 

Not at all, I must have mis-read something (or they released 9 versions in the last week :) ). I'll update the top comment to avoid confusion.

> I forked meteor and merged devel onto mongo-3.2 there was a couple small changes. I plan to keep merging devel frequently so I don't end up much behind. Do you think this is a good strategy?

No reason not to but I wouldn't expect many mongo-touching changes to devel to land.

**Comment by tmeasday** [comment]
That `httpcall -errors` test certainly shouldn't depend on mongo in anyway, so something odd is going on. Do you see the error on latest `devel`?

**Comment by Fabs** [comment]
> That httpcall - errors test certainly shouldn't depend on mongo in any way, so something odd is going on. Do you see the error on latest devel?

I took the time now to read the test carefully, I think maybe I know what the issue could be. I have OSX Server, and the address http://0.0.0.0/ resolves an webpage (apache is running, port 80).

So the assumption behind the test `HTTP.call("GET", "http://0.0.0.0/", expect(unknownServerCallback));` does not hold on my machine, and this the test fail. When I turn the server off, then the test stop failing. http://0.0.0.0:3000/ also resolves to my meteor server. Different from a vanilla mac, I can say I have dnsmasq with some confs for Docker and other machines at my network, even localhost, but nothing explicitly targeting 0.0.0.0

Also, I am at `devel` and there are few small differences from the past 6 hours (https://github.com/Fabs/meteor/pull/1) but none I believe relevant to the failing test.

**Comment by Fabs** [comment]
On a side note, already used something from @dgtlife to fix finding the primary replica set upon connection :). Thanks!

**Comment by tmeasday** [comment]
Whoa, `0.0.0.0` wasn't a great choice! Is there a known "bad" IP address? We should change the test to use one that doesn't resolve to localhost (I'm surprised this hasn't come up before). Perhaps `240.0.0.0` is a better choice? https://tools.ietf.org/html/rfc5735

Re: #4970 -- that test includes a timeout. I wonder if that's what's

**Comment by tmeasday** [comment]
PS I am on leave from tomorrow until Tuesday but @benjamn will jump in and help out with any other questions.

**Comment by abernix** [comment]
Opened #7038 for that failed test and I'll follow through on it.  Nice find @Fabs.  Continue as you were! :)

**Comment by Fabs** [comment]
About #4970, I am thinking about playing with it once I have the mongo done. Is that alright? 

Thanks @abernix. Wish you a good time @tmeasday, thanks for the help you gave already.

Just an update, so far, `using MONGO_URL without` it is looking good. There are about 60 tests broken on mongo-livedata, and a couple more here and there (accounts...). Most of the broken tests are about the results that come from the driver, for instance when Meteor.update is expected to return the number of modified documents, it is now returning an object with a lot of extra stuff...

**Comment by benjamn** [comment]
👋  folks

**Comment by guilhermedecampo** [comment]
This is pretty awesome! 👍

**Comment by Fabs** [comment]
Update time :). 

The upsert method, specially [`simulateUpsertWithInsertedId`](https://github.com/meteor/meteor/blob/abd574f38008b45f5e2a6bc322b10bcdde44763a/packages/mongo/mongo_driver.js#L614) was causing many failing tests. The current `node-mongodb-native` update method works with `$setOnInsert` and offers now, amidst its results, a list of `upserted` documents, where we can find the ID of the inserted document (which I believe was one of the restrictions preventing using a single update for upsert). I also preserved the `NUM_OPTIMISTIC_TRIES` behaviour, although I am not sure if it is still necessary.

Anyways, I am still dwelling in some corner cases, but I am confident I will be able to get rid of the previous implementation of upsert (with the `doUpdate` and `doConditionalInsert`) in favor of one with a single update and upsert:true. :) 

**Got 20 tests working today, 40 more to go.**

**Comment by dgtlife** [comment]
From #5809, this may also be relevant [https://jira.mongodb.org/browse/SERVER-14322](https://jira.mongodb.org/browse/SERVER-14322)

**Comment by Fabs** [comment]
More Updates. 

I am now 5 failing tests to have it working without oplog and using upsert. All my troubles are with the _id field, but I think I am getting close. I believe by the end of this week I will have it working for this particular case.

If anybody has any ideas, one problem I am dealing is setting the "right" ID, on STRING id type collections when the user supply an upsert (or update with, upsert:true) without any modifier operators (`col.upsert({foo: 1},{foo: 2, bar: 3})` for instance). If the insert happens, mongo will generate an ObjectId type ID. 

I try using $setOnInsert to add a String ID (and $set for the rest of the update document). This fix the insert but brakes the update, when the expected behaviour would be replacing the entire document {$set does not do that}. 

Also I think I break some behaviour that does not have failing tests so I will take time to double check it and add the tests.

**Comment by toszter** [comment]
@Fabs --- I see you're "personal" mongo-3.2 branch is 42 commits (at the time of this writing) ahead of the "official" mongo-3.2 branch.  

Is it safe to say---despite not having the last test completed---that we can clone your "personal" mongo-3.2 in order to begin experimenting with MongoDB 3.2 behind Meteor?  

Will this become the PR for 1.4?

I went though the entire gitlog on your branch. What's left to fix after tests are complete and passed?

Thank you for moving this forward! Hats off to you.

**Comment by Fabs** [comment]
Hey @toszter. By all means you can play with it, with lucky you would spot some problems I missed at this stage :). 

To give you and everybody a state of affairs :D. 

If you `do not use oplog` I think the only problem you will find is that the ids returned by an upsert (when the insert happens) are of the wrong object type (the driver ObjectId instead of Mongo.ObjectID). There is another thing regarding upsert, but I don't think it breaks anything, although I will have to fix some optimistic writing since it looks like it is the way other writes go about it (I talked in more detail about my upsert difficulties yesterday, on a comment above). Anyway, I am planning to fix this two points, clean some debugging, organize, comment a little bit more, refactor and have it done by the end of this week (maybe already with a PR to meteor/mongo-3.2). Then I would say its a first milestone reached. 

From there I will go to the oplog. If you `want to use oplog` today, I do not know the state of affairs. So far, I did not touch anything oplog specific but I was not even being able to connect to oplog, on 3.2 just by switching the driver.

I think for it to be 1.4 ready. I still have to switch the`dev-bundle` to have mongo 3.2, and leave some nice messages for people about the transition (I am nor sure weather there will have breaking changes). I did not discussed anything with the guys from MDG regarding testing it more profusely. I was thinking that after I have the PR done we could have some sort of `ci` running the code against different versions of mongodb.... I'd love to have it done for the next version, that is sure. 

There is also a bug with WiredTiger that looks pretty bad, and I still want to consider how bad it can affect us (https://jira.mongodb.org/browse/SERVER-14322).

It is my first time around the source code of meteor, I am doing my best to have it quickly done, and I would appreciate any feedback or help in order to get it better/faster :). It's a challenging change I am having lots of fun.

**Comment by dgtlife** [comment]
@Fabs I am able to use `collection.update`, `collection.update` with `upsert: true`, and `collection.upsert` (with API conformance per Meteor docs and without any errors) using the modifications I just pushed to my [repo](https://github.com/dgtlife/meteor-mongo-for-3.2.x).

**Comment by Fabs** [comment]
@dgtlife. I took a look at your code. I pretty much went the same strategy. To fix the result after the methods. I tried to maintain changes on the driver instead of collection so in the future if the driver changes the collection does not care. 

Did u run against the package tests? For me, there is just few corner cases now where it does not work. 

Anyway thanks very much. I will look what u did with attention and see if it can help me.

**Comment by Fabs** [comment]
Good news. All tests `non oplog` for the `mongo` package are passing \0/. I forgot though, two broken tests on the `accounts` package. I will fix then and open a PR with it to `mongo-3.2` probably today, maybe tomorrow.

I took a step back on the switching to native upsert (for the moment). But I will try again :).

I have open two PR on my branch [one for @dgtlife code agains devel](https://github.com/Fabs/meteor/pull/4) and other for my first atempt [with the upsert agains devel](https://github.com/Fabs/meteor/pull/2). so it helps compare the code.

The branch I will PR is the new [Fabs/mongo-3.2-fix-only ](https://github.com/Fabs/meteor/pull/5)

**Comment by Fabs** [comment]
There is a breaking change that I'd like to discuss, where maybe @tmeasday and @benjamn know a way to do about it. 

The accounts package was broken because it relied on `err` object return by the collection to have an `err` field containing the error message. This field is now called `errmsg` on the new driver API. 

Should I patch all the `err` to fix it and be backwards compatible? I imagine it can break people relying on the error message. Should I have both `err` and `errmsg`? I honestly don't have an strong opinion (nor a great idea) here, so I would appreciate some input.

**Comment by tmeasday** [comment]
Hey @Fabs. I would suspect it might be better to just update Accounts. Obviously others might be relying on the same behavior but we probably can't keep layering on back-compat forever.

**Comment by dgtlife** [comment]
@Fabs I misunderstood your statement about "ids returned by an upsert ...". I thought you were still having difficulty with the `writeResult`.

In my hack I'm just fixing the errors where they are thrown based on running tests in my app. I am not using the package tests (definitely more cases), nor necessarily observing architectural aesthetics ;-)

**Comment by tmeasday** [comment]
PS @dgtlife I suspect you may be interested in: https://github.com/meteor/meteor/issues/6958 -- we are thinking of a different API to setting all the various connection options

**Comment by dgtlife** [comment]
@tmeasday OK. I'll track it.

**Comment by Fabs** [comment]
> @Fabs I misunderstood your statement about "ids returned by an upsert ...". I thought you were still having difficulty with the writeResult.

It was great you uploaded that code anyway. Because I finally realised that it was 99% result errors, and so I took a step back and fixed everything on top of the previous upsert implementation. Anyway, I think what you did is pretty close to what I did, and your code definitely has helped a lot.

**Comment by abernix** [comment]
So this is purely as food for thought and I know this only affects dev environment, but I thought I'd bring up the possibility of not changing the storage engine by default.  (Not for personal reasons either!).  

Mongo 3.2 still has a lot of very valuable updates even if this change doesn't automatically make the big storage engine jump to WiredTiger.  It's likely that many production deployments will continue to use MMAPv1 for the foreseeable future and this driver update will still allow that.  But, this might be a good stage to start recommending that production users consider making the change.

Would it be at all reasonable to keep _existing_ dev environments on MMAPv1 with a suggestion banner about the need to upgrade and instructions?  It might give them the message (and exact instructions) that something will need to change in production too and help them understand how to do it.

Then any new `meteor create` projects could use WiredTiger with no banner.  (DB engine can be detected with `db.serverStatus().storageEngine`)

(Also, maybe `mupx` needs a migration tool?)

**Comment by tmeasday** [comment]
I think your idea has merit @abernix. Let me think about it.

**Comment by abernix** [comment]
Not trying to feature creep but this but #4947 might be a reasonable to fix if someone was already digging around in Oplog Observer code.

**Comment by Fabs** [comment]
> Not trying to feature creep but this but #4947 might be a reasonable to fix if someone was already digging around in Oplog Observer code.

Soooooo. I got oplog connecting :). `S: entry skipping fails ` and no tests run after it. I was expecting a more catastrophic explosion, but I think I wont be much delayed on get it working. Then I can look #4947. I dunno if @tmeasday did anything more then confirming.

**Comment by tmeasday** [comment]
Oh, I just added the confirmed label to this feature because I realised I wasn't following the process :)

**Comment by tmeasday** [comment]
@Fabs I think #4947 is very much a nice-to-have.

**Comment by Fabs** [comment]
> @Fabs I think #4947 is very much a nice-to-have. (By which I mean, only work on it if it's easy or you especially want to).

All right! I will certainly focus on what is must haves for 1.4. When they are done, I will see if I still can address the next things.

**Comment by abernix** [comment]
It would _seem_ that #4436 will be getting resolved through this upgrade as the [`transformResult` method will use `nModified`](https://github.com/Fabs/meteor/blob/fd114d673b13c3b9541475d1aa9455d853154659/packages/mongo/mongo_driver.js#L604-L629).

I'm not certain, but it would seem that Meteor was previously using the `result` from `update` which was records _matched_, **not** records _modified_.

If this is the case, this should be mentioned very clearly in the `History.md` as a breaking change.

**Comment by tmeasday** [comment]
@Fabs I've been pondering @abernix's comment above.

At first I thought it made sense to just stick with changing the `.update()` API to return `nModified` rather than `nMatched` as suggested in #4436, and document the change.

But re-reading the issue and thinking about it more, I have some objections to that:

 1. It seems more than anything to be a *documentation* issue---as @davidecantinit [pointed out](https://github.com/meteor/meteor/issues/4436#issuecomment-167103377) there are times where both return values are useful---@mitar's main point seems to be the behaviour doesn't match his expectation from the docs (and the solution seems easy, just change the docs to say "number matched" rather than  the vague "number affected").

 2. We'd have to change the API of minimongo too to be consistent.

 3. It's something people could very easily miss.

There just doesn't seem to be all that a compelling case for sticking with changing it. I'd love to hear your thoughts, but I'm thinking we should change it back to `nMatched` (and add a test for it).

**Comment by mitar** [comment]
I have never ever needed number of matched documents, but I have always needed number of changed documents. Can somebody tell me use cases where first would be preferred over others. (Let's imagine that we are designing this from scratch and that we can decide on one value to be returned.)

We also have documentation so that we have an "excuse" to fixing this in a backwards incompatible manner. But we should add some switch to return the whole object for those who really need previous behavior. But I really do not know when you would need number of matched documents.

**Comment by Fabs** [comment]
I have not changed any of the tests during the driver changed and I returned nModified (or 1 for insertions) for all update tests. So, I believe either the documentation was always wrong, or there is no test where `nModified != nMatched` and an explicit check on numberAffected expecting it to be nMatched happens (I will make sure there is a test for this behaviour anyway). 

My preference would be to do something in the direction of @davidecantini proposal, returning at least two values `nMatched` and `nModified` (on the insert, there is the insertedId) and get rid of numberAffected. I agree with @mitar that the `nModified` seems to *me* the most useful result, I never ever even noticed that behaviour before, I would always just go `oh, the result of update has to be the number of changed things`, but I also can see when other developers reasoning differently specially with the current documentation. 

One feature I think we are in a good position to add now though, is returning the whole `rawResult` by specifying an option on the operation (update/insert/remove...) `fullResult: true` or `result: 'full'` or ... . The idea here is that I will always have a `meteor style` result as default, but whenever I want the internal driver format (that could even be rethinkDB or mysql or whatever driver) I can have a more complete result without resorting to rawCollection operations.

Another more simple idea, would be returning the internal result object as a property of the meteor result, all the time.

**Comment by Fabs** [comment]
Guys, am I missing something? I went to the docs, and they read:

> Modify one or more documents in the collection, or insert one if no matching documents were found. Returns an object with keys numberAffected (the number of documents modified) and insertedId (the unique _id of the document that was inserted, if any).

Was that updated?

**Comment by tmeasday** [comment]
> Can somebody tell me use cases where first would be preferred over others

This comment explains some examples where both are useful: https://github.com/meteor/meteor/issues/4436#issuecomment-167103377. Ultimately, this is clearly why MongoDB changed the wire protocol.

> Let's imagine that we are designing this from scratch and that we can decide on one value to be returned

Unfortunately we don't live in a world where we can design things in a white room. Back-compat is a real thing. If I was designing it from scratch, I'd return the full `{nModified, nMatched, nUpserted}` document that Mongo does over the wire.

However, I'm wary of making that change to the default API, because if anyone has code of the form `if (! X.update(...))` right now, changing the API to always return an object will definitely break it.

>  I believe either the documentation was always wrong, or there is no test

The document was indeed always wrong[1], and I think you are right that there is no test.

> One feature I think we are in a good position to add now though, is returning the whole rawResult by specifying an option

I think this is the right approach. It should be easy to send a PR to do this after 1.4

> Guys, am I missing something?

You are looking at the *upsert* docs. The [*update*](http://docs.meteor.com/api/collections.html#Mongo-Collection-update) docs say: "Returns the number of affected documents."

[1] Or at best confusing.

**Comment by tmeasday** [comment]
I just don't think there's a strong enough case to be made to break back-compat (even though I agree most people probably would never had made the distinction between `nModified` and `nMatched`, esp given the docs implied the first).

I definitely think we should add a test to avoid unintentionally changing this in future (and to ensure consistency between the Mongo Driver and Minimongo).

**Comment by mitar** [comment]
Hm, but to me it is a bug. I was programming against documentation and I reported a bug that Meteor is not returning correct value. This change will always be incompatible for some users. Or those who read documentation, or those who ignored it. The question is then into which direction we want to move. Which set of users we want to inconvenience?

**Comment by tmeasday** [comment]
I don't think it's fair to say that it's "breaking compatibility" if people mis-understood the API because of poor documentation and *we don't change it*.

**Comment by mitar** [comment]
I had in one my code for 2 years this bug because I read documentation and do not always check implementation because I trust code. I suspect that others have similar code based on documentation. For them this is a breaking change.

**Comment by tmeasday** [comment]
How can it be a breaking change when we aren't changing anything? You've lost me..

**Comment by mitar** [comment]
It is a breaking change because for those who implemented code based on documentation have a hidden bug. I had such bug for two years and I had very nasty issues with it: my code went into an infinite loop blocking the whole server. Of course it was hard to find it. Because it happened only when there was a race condition that document has already the same values as I wanted to update to. Tricky to debug. Probably my case is an extreme case, but still.

Anyway, I do not really case. I fixed it for myself and it works. I just think more people program against documentation than program against implementation.

**Comment by davidecantini** [comment]
Hi guys,
Considering the two options: 

a) change the documentation to reflect the current (and past) implementation; 
b) change the implementation, to reflect the current (and past) documentation;

The question should be: which one has less chances of breaking existing codes?

Any real code/app gets tested and it's the implementation, in the end, that causes your tests to pass or fail. So, while it's true that one should follow the documentation, it's also true that as soon as your tests fail (because you expected nModified and actually got nMatched) you'd surely have to follow the implementation (if you want your tests to pass). 

That said, it appears reasonable to say that option a) has less chances of breaking existing codes than option b), and, so, a) should be the way to go. If so, would the following be something we can agree on?

1. Replace "Returns the number of affected documents." with "Returns the number of matched documents" to reflect the actual behavior of [Meteor.Update](http://docs.meteor.com/api/collections.html#Mongo-Collection-update) and add a "Possible breaking change" note, even though in most cases, if not all, there won't be anything broken (for the reasons mentioned above).
2. Add an option to return the whole raw result (as mentioned by [Fabs](https://github.com/meteor/meteor/issues/6957#issuecomment-223927058)) which can give full control, for those who need it.

I hope this helps, thanks.

**Comment by mitar** [comment]
> it's also true that as soon as your tests fail (because you expected nModified and actually got nMatched) you'd surely have to follow the implementation (if you want your tests to pass). 

Even Meteor do not have tests for this edge case. And this is pretty tricky to test because it might depend on a race condition.

**Comment by davidecantini** [comment]
@mitar, do you believe that [option b)](https://github.com/meteor/meteor/issues/6957#issuecomment-224254318) would have less chances of breaking existing codes than option a)?

**Comment by mitar** [comment]
I think so, but only because of my sample size of 1 (me). :-) Anyway, I will leave to others to decide this. I have spoken too much on this topic.

**Comment by davidecantini** [comment]
@mitar, it's fine to have opinions, regardless of the sample size :) Here is a simple case from a real app:
```
let n = Cards.update({_id: card._id}, {$set: card});
if (n === 0) {
    // No match: the document may not exist. Throw an error and show a UI error msg
     throw new Meteor.Error(...);
}
```
As you can see, that is a case where a user opens the UI to update an item and saves it without modifying anything. It's something quite common and easy to test. Well, if you go for option b) then each and every existing code similar to that would get broken.
So, all in all, [option a)](https://github.com/meteor/meteor/issues/6957#issuecomment-224254318) appears to be the lesser of two evils here.

**Comment by Fabs** [comment]
So. I Made numberAffected return `nMatched`m the backwards compatible way :). 
https://github.com/meteor/meteor/pull/7190

I hope once those modifications of the driver hit 1.4, somebody (probably me, but anyone could ;)) will make the `rawResult` approach we've been discussing. 

About the docs. What is the process to fix them (if no one is planning to do that I volunteer)?

@mitar , do you think once we have added the `rawResult` proposed feature, it will solve the problem for you?

**Comment by mitar** [comment]
Yea, it is OK.

**Comment by tmeasday** [comment]
> About the docs. What is the process to fix them (if no one is planning to do that I volunteer)?

https://github.com/meteor/meteor/pull/7163/commits/e2904d9c7c25064faa0aab4893b077beb248f652

**Comment by jedwards1211** [comment]
@tmeasday the fork of `mongo` is for dev only, right?

**Comment by tmeasday** [comment]
Actually we don't use a fork any more. But it was just for dev, yeah.

**Comment by benjamn** [comment]
The changes associated with this issue have been merged into [`release-1.4`](https://github.com/meteor/meteor/pull/7218), and the corresponding [PR](https://github.com/meteor/meteor/pull/7163) is closed, so I think we can close this issue now. Thanks @tmeasday and @fabs and everyone else for making this happen!
